### PR TITLE
use avro DataFileReader to read avro container files

### LINKED ISSUE #23213: [Feature Request]: AvroSource should use the standard CodecFactory to decompress blocks
### What would you like to happen?

Currently AvroSource has its own hand-rolled avro parser, given the class is over 6 years old I assume this is due to it being written before Avro had DataFileReader, etc, with good support for partially reading an avro container file.

One of the downsides of this is that it doesn't share the same CodecFactory infrastructure as avro proper does, so users can't plug in custom CodecFactory instances using `CodecFactory.addCodec`.

It should be possible to refactor AvroSource to use DataFileReader rather than its own handmade parser, and in doing so would automatically get the benefit of using the normal CodecFactory infrastructure.

### Issue Priority

Priority: 3

### Issue Component

Component: io-java-avro

### PR DESCRIPTION
**Pull request by steveniemitz**
This fixes #23213

This refactors AvroSource to use the standard DataFileReader from Avro.  I don't really have the context around why a custom file parser was written in beam for this, but the code is so old I assume that DataFileReader couldn't properly support reading from arbitrary positions in a container file when it was written.

R: @iemejia  (you were the last person other than myself to substantially change this class)

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by steveniemitz** [comment]
Run Java_PVR_Flink_Docker PreCommit

**Comment by steveniemitz** [comment]
Run Kotlin_Examples PreCommit

**Comment by steveniemitz** [comment]
ping @iemejia

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroSource.java [hunk 1]
```diff
@@ -561,77 +553,24 @@ private Object readResolve() throws ObjectStreamException {
...
-          return new XZCompressorInputStream(byteStream);
-        case DataFileConstants.BZIP2_CODEC:
-          return new BZip2CompressorInputStream(byteStream);
-        case DataFileConstants.NULL_CODEC:
-          return byteStream;
-        default:
-          throw new IllegalArgumentException("Unsupported codec: " + codec);
-      }
-    }
+    private final long numRecordsInBlock;
```
Comment (iemejia) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroSource.java:567]: Much better name 👍

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroSource.java [hunk 2]
```diff
@@ -669,13 +608,49 @@ public double getFractionOfBlockConsumed() {
    */
   @Experimental(Kind.SOURCE_SINK)
   public static class AvroReader<T> extends BlockBasedReader<T> {
-    // Initialized in startReading.
-    private @Nullable AvroMetadata metadata;
+
+    private static class SeekableChannelInput implements SeekableInput {
```
Comment (iemejia) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroSource.java:612]: A pity this one does not exist in Avro

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroSource.java [hunk 3]
```diff
@@ -717,71 +678,25 @@ public synchronized AvroSource<T> getCurrentSource() {
     //
     // Postcondition: same as above, but for the new current (formerly next) block.
     @Override
-    public boolean readNextBlock() throws IOException {
-      long startOfNextBlock;
-      synchronized (progressLock) {
```
Comment (iemejia) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroSource.java:722]: 👍

**Review by iemejia — APPROVED** [review]
LGTM Really nice simplification/reuse. (My excuses for the delay to do the review).

**Pull request merged**